pkg/reconciler/managed/api.go

```diff
@@ -130,6 +143,23 @@ func NewAPISimpleReferenceResolver(c client.Client) *APISimpleReferenceResolver
 	return &APISimpleReferenceResolver{client: c}
 }
 
+func prepareJSONMerge(existing, resolved runtime.Object) ([]byte, error) {
+	// restore the to be replaced GVK so that the existing object is
+	// not modified by this function.
+	defer existing.GetObjectKind().SetGroupVersionKind(existing.GetObjectKind().GroupVersionKind())
+	existing.GetObjectKind().SetGroupVersionKind(schema.GroupVersionKind{})
```

Comment:
Sorry, I couldn't get the reasoning behind these two lines. What happens if we didn't have both?

Reply:
Hi @turkenh,
The deferred resetting in line 149 is currently not needed and we can remove it. I felt like the function is actually not expected to modify the object `existing` which is generally a pointer type and instead of deep copying and modifying the copy, I preferred the much lighter set/reset pattern. Currently, the caller does not use `existing` after it calls this function but if we attempt to do so assuming `existing` is unmodified, that assumption now holds because we reset the GVK to its original value.

In line 150, we set the GVK to its zero value because in the calculated patch document we need the type meta as in:
```json
{
  "apiVersion": "ec2.aws.upbound.io/v1beta1",
  "kind": "Subnet",
...
```

I will leave a comment explaining this.

Reply:
I see, thanks! 
Another idea could be setting those two fields over the generated patch to make it more obvious.